### ISSUE TITLE
Sitemap.xml: lastmod timestamp can contain invalid dates

### ISSUE THREAD

**Issue by hostep** [magento/magento2#9151]
### Preconditions
1. Magento CE 2.1.5

### Steps to reproduce
1. Have a category in the database where the `created_at` timestamp is a valid date, and the `updated_at` timestamp is `0000-00-00 00:00:00`
I don't think you can manage this by manually creating a category in the backend of Magento, but we managed to have some categories like this by importing them using a custom PHP script.
2. Generate a sitemap.xml file
3. Look at the generated sitemap.xml file for that category.

### Expected result
1. The `lastmod` value in the xml should contain the `created_at` timestamp

### Actual result
1. The `lastmod` value in the xml looks like this:
```xml
<lastmod>-001-11-30T00:00:00+00:00</lastmod>
```
The negative value is most likely caused by our timezone setting: `Europe/Brussels`.

### Possible solution
```diff
diff --git a/app/code/Magento/Sitemap/Model/Sitemap.php b/app/code/Magento/Sitemap/Model/Sitemap.php
index 3b6f646f9a5..434393ae63a 100644
--- a/app/code/Magento/Sitemap/Model/Sitemap.php
+++ b/app/code/Magento/Sitemap/Model/Sitemap.php
@@ -341,9 +341,10 @@ class Sitemap extends \Magento\Framework\Model\AbstractModel
             $changefreq = $sitemapItem->getChangefreq();
             $priority = $sitemapItem->getPriority();
             foreach ($sitemapItem->getCollection() as $item) {
+                $lastmod = $item->getUpdatedAt() === '0000-00-00 00:00:00' ? $item->getCreatedAt() : $item->getUpdatedAt();
                 $xml = $this->_getSitemapRow(
                     $item->getUrl(),
-                    $item->getUpdatedAt(),
+                    $lastmod,
                     $changefreq,
                     $priority,
                     $item->getImages()
```

### Discussion
I'm not sure if this is an actual bug in Magento, due to the fact that we inserted some data in the database in a possibly incorrect way.

Also: what should happen when the `created_at` timestamp contains '0000-00-00 00:00:00'? In that case the `lastmod` field shouldn't appear I assume, since it is optional: https://www.sitemaps.org/protocol.html#xmlTagDefinitions

If somebody could let me know what the best solution would be in this case, I can create a pull request for this.
If it is decided that this isn't an actual bug, then we should fix our custom import code I guess.

Please let me know, thanks! :)

**Comment by korostii** [comment]
IMHO setting `updated_at `to a meaningful value during the import seems most logical, regardless of the outcome of this issue.

Back to the issue in question, a negative date value is something new :rofl:

**Comment by magento-engcom-team** [comment]
@hostep, thank you for your report.
The issue is already fixed in develop branch
But we will consider to backport the fix to patch releases

**Comment by hostep** [comment]
@magento-engcom-team: can you point me to the commit which fixes this, since I don't see it?
The last change to the last `<lastmod>` in the `Magento\Sitemap\Model\Sitemap` class was done 4 years ago ...

I think you are talking about issue 10598 which is not the same as this one :)

**Comment by okorshenko** [comment]
The issue has been fixed in 2.2-develop branch. Will be available with 2.2.2 release

**Comment by okorshenko** [comment]
Hi @hostep. Thank you for your report. 
The issue has been fixed in magento-engcom/magento2ce#1353 by @magento-engcom-team in 2.3-develop branch
Related commit(s):
 - 299109f1265fa15eeb3067e22c2638ff5907205e
 - 57318bf03bf6fe7f84b2956ffb5ecc1332bf41b3
 - 41cd961a217d30c4b7b051a775902c4fb40b84ea
 - 076b90398852478d93d1444f936ac483b4640a82
 - 435a6c200cd548b8765e5bc73b0458039a58645d
 - 5d174b873f8ba369d6504fb9d7ef30296816e2b1
 - a47b10dce854a2d21ead6b362b86ca3ddf595386
 - ab48cb176830ddab69fd1617cae96d76094ab167
 - 46be4c7ea59f6cfb9884f2bc5875245b1960edc6
 - eafc6f0a473547542c3ef80352dcd2be7128da87
 - 63b9e909f1173a51374d4cdfc68249baa66f0007
 - 2edfe9a30bb64f0a7e102bb335a89658b627bfad
 - ef5ddffb43a735627d09535b62fac2a41804a96a
 - 4ada797ece84c2f5eb55499e8bc3f75e0c7ef245
 - c5d9a019b0d51a51b452fd99b527e3b894874be2
 - 72e9e0732b98ea5b8f47bddc781dc47c688c3c58
 - 215180de0a3f6c4a16e02205732bfe8670293ad2
 - 685d50625b68dbb960988a6bfbbfc14455a6cb9e

The fix will be available with the upcoming 2.3.0 release.
